File: app/src/main/java/com/shetj/diyalbume/pipiti/localMusic/LocalMusicActivity.java
package com.shetj.diyalbume.pipiti.localMusic;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import com.alibaba.android.arouter.facade.annotation.Route;
import com.chad.library.adapter.base.BaseQuickAdapter;
import com.jakewharton.rxbinding2.view.RxView;
import com.shetj.diyalbume.R;
import com.shetj.diyalbume.pipiti.media.MediaPlayerUtils;
import com.shetj.diyalbume.pipiti.media.SimPlayerListener;

import java.util.ArrayList;
import java.util.List;

import me.shetj.base.base.BaseActivity;
import me.shetj.base.base.BaseMessage;
import me.shetj.base.tools.app.ArmsUtils;
import me.shetj.base.tools.json.GsonKit;
import timber.log.Timber;

/**
 * 本地音乐
 *
 * @author Administrator
 */
@Route(path = "/shetj/LocalMusicActivity")
public class LocalMusicActivity extends BaseActivity<LocalMusicPresenter> {

	private RecyclerView mIRecyclerView;
	private MusicSelectAdapter mAdapter;
	private MediaPlayerUtils mediaUtils;
	private int currentPosition = 0;
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_local_music);
		ArmsUtils.statuInScreen(this,true);
		initView();
		initData();
	}

	@Override
	protected void initView() {
		TextView title = findViewById(R.id.toolbar_title);
		title.setText("本地音乐");
		RxView.clicks( findViewById(R.id.toolbar_back)).subscribe(o -> onBackPressed());
		mediaUtils = new MediaPlayerUtils();
		mIRecyclerView = (RecyclerView) findViewById(R.id.IRecyclerView);
		ArmsUtils.configRecycleView(mIRecyclerView,new LinearLayoutManager(this));
		mAdapter = new MusicSelectAdapter(new ArrayList<>());
		mAdapter.bindToRecyclerView(mIRecyclerView);
		mAdapter.setOnItemClickListener(new BaseQuickAdapter.OnItemClickListener() {
			@Override
			public void onItemClick(BaseQuickAdapter adapter, View view, int position) {
				ArmsUtils.makeText(GsonKit.objectToJson(mAdapter.getItem(position)));
			}
		});
		mAdapter.setOnItemChildClickListener(new BaseQuickAdapter.OnItemChildClickListener() {
			@Override
			public void onItemChildClick(BaseQuickAdapter adapter, View view, int position) {
				currentPosition = position;
				mediaUtils.playOrStop(mAdapter.getItem(position).url,new SimPlayerListener(){
					@Override
					public void onStart(String url) {
						super.onStart(url);
						ArmsUtils.makeText(String.format("试听%s", mAdapter.getItem(position).name));
					}

					@Override
					public void onPause() {
						super.onPause();
						ArmsUtils.makeText( "暂停" );
					}

					@Override
					public void onProgress(int current, int size) {
						super.onProgress(current, size);
						Timber.i(current+"/"+size);
					}

					@Override
					public boolean isNext(MediaPlayerUtils mp) {
						currentPosition++;
						if (currentPosition < mAdapter.getItemCount()) {
							mp.playOrStop(mAdapter.getItem(currentPosition).url, this);
							return true;
						}else {
							return false;
						}
					}

					@Override
					public void onCompletion() {
						super.onCompletion();
						ArmsUtils.makeText("播放结束"+mAdapter.getItem(currentPosition).url);
					}
				});
			}
		});
	}

	@Override
	protected void initData() {
		mPresenter = new LocalMusicPresenter(this);
		mPresenter.initMusic();
	}

	@Override
	public void updateView(BaseMessage message) {
		super.updateView(message);
		switch (message.type){
			case 1:
				mAdapter.setNewData((List<Music>) message.obj);
				break;
			default:
				break;
		}
	}

	@Override
	protected void onPause() {
		super.onPause();
		if (null != mediaUtils) {
			mediaUtils.onPause();
		}
	}

	@Override
	protected void onResume() {
		super.onResume();
		if (null != mediaUtils) {
			mediaUtils.onResume();
		}
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		if (null != mediaUtils) {
			mediaUtils.onDestroy();
		}
	}
}

File: app/src/main/java/com/shetj/diyalbume/pipiti/localMusic/LocalMusicActivity.kt
package com.shetj.diyalbume.pipiti.localMusic

import android.os.Bundle
import android.os.Message
import android.view.View
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.alibaba.android.arouter.facade.annotation.Route
import com.shetj.diyalbume.R
import com.shetj.diyalbume.pipiti.media.MediaPlayerUtils
import com.shetj.diyalbume.pipiti.media.SimPlayerListener
import me.shetj.base.base.BaseActivity
import me.shetj.base.tools.app.ArmsUtils
import me.shetj.base.tools.app.ArmsUtils.Companion.statuInScreen
import me.shetj.base.tools.json.GsonKit
import timber.log.Timber
import java.util.*

/**
 * 本地音乐
 *
 * @author Administrator
 */
@Route(path = "/shetj/LocalMusicActivity")
class LocalMusicActivity : BaseActivity<LocalMusicPresenter>() {

    private var mIRecyclerView: RecyclerView? = null
    private var mAdapter: MusicSelectAdapter? = null
    private var mediaUtils: MediaPlayerUtils? = null
    private var currentPosition = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_local_music)
        statuInScreen( true)
    }

    override fun initView() {
        val title = findViewById<TextView>(R.id.toolbar_title)
        title.text = "本地音乐"
        findViewById<View>(R.id.toolbar_back).setOnClickListener { onBackPressed() }
        mediaUtils = MediaPlayerUtils()
        mIRecyclerView = findViewById<View>(R.id.IRecyclerView) as RecyclerView
        mAdapter = MusicSelectAdapter(ArrayList())
        mIRecyclerView?.adapter = mAdapter
        mAdapter!!.setOnItemClickListener { _, _, position -> ArmsUtils.makeText(GsonKit.objectToJson(mAdapter!!.getItem(position)!!)!!) }
        mAdapter!!.setOnItemChildClickListener { _, _, position ->
            currentPosition = position
            mAdapter!!.getItem(position)!!.url?.let {
                mediaUtils!!.playOrStop(it, object : SimPlayerListener() {
                    override fun onStart(url: String) {
                        super.onStart(url)
                        ArmsUtils.makeText(String.format("试听%s", mAdapter!!.getItem(position)!!.name))
                    }

                    override fun onPause() {
                        super.onPause()
                        ArmsUtils.makeText("暂停")
                    }

                    override fun onProgress(current: Int, size: Int) {
                        super.onProgress(current, size)
                        Timber.i("$current/$size")
                    }

                    override fun isNext(mp: MediaPlayerUtils): Boolean {
                        currentPosition++
                        return if (currentPosition < mAdapter!!.itemCount) {
                            mAdapter!!.getItem(currentPosition)!!.url?.let { it1 -> mp.playOrStop(it1, this) }
                            true
                        } else {
                            false
                        }
                    }

                    override fun onCompletion() {
                        super.onCompletion()
                        ArmsUtils.makeText("播放结束" + mAdapter!!.getItem(currentPosition)!!.url)
                    }
                })
            }
        }
    }

    override fun initData() {
        mPresenter = LocalMusicPresenter(this)
        mPresenter!!.initMusic()
    }

    override fun updateView(message: Message) {
        super.updateView(message)
        when (message.what) {
            1 -> mAdapter!!.setNewData(message.obj as MutableList<Music>)
            else -> {
            }
        }
        message.recycle()//释放
    }

    override fun onPause() {
        super.onPause()
        if (null != mediaUtils) {
            mediaUtils!!.onPause()
        }
    }

    override fun onResume() {
        super.onResume()
        if (null != mediaUtils) {
            mediaUtils!!.onResume()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (null != mediaUtils) {
            mediaUtils!!.onDestroy()
        }
    }
}
